File: cloud/lib/barkpark_cloud/verify.ex
defmodule BarkparkCloud.Verify do
  @moduledoc """
  On-demand golden-path VERIFY (C8 / charter D53) — readiness as a re-runnable
  proof.

  The Go provisioner runs a birth-certificate probe gate BEFORE it ever declares
  a box `ready` (`internal/provisioner/verify.go`, C2/D45): a box that would 500
  the moment its session/auth stack is exercised (the #957 32-byte
  SECRET_KEY_BASE class) is never called live. This module is the SAME suite,
  re-issuable on demand from the console (`POST /v1/barkparks/:id/verify`) and
  the CLI (`bp cloud verify`) — a ping tells you the box answers TCP; this tells
  you the golden path still works.

  ## The suite (one vocabulary, two executors — the D32 discipline)

  The probe VOCABULARY (names / labels / pass-rules) is committed as a shared
  fixture, `priv/static/__fixtures__/verify_probes.json`, asserted from BOTH
  this executor (`probes/0` deep-equals it) AND the Go gate's compiled probe
  order (`internal/provisioner/verify_fixture_test.go`). Two hand-written probe
  lists WILL drift; one fixture with two asserters can't.

    * `verify.api`    GET  `/v1/capabilities` (stored admin token) → `200`
    * `verify.login`  POST `/v1/auth/login` (deliberately-wrong sentinel creds)
      → status `< 500` — a 401/422 proves the whole request→session→auth
      pipeline ran and cleanly rejected the creds; ANY 5xx is the #957
      dead-on-arrival class.
    * `verify.studio` GET `/studio`, following the scoped-login redirect ≤3 hops
      → final status `< 500`.

  ## Token custody (D46)

  The per-instance admin bearer is decrypted server-side via
  `Registry.reveal_admin_token/1` and sent ONLY on the `verify.api` probe's
  `Authorization` header. It is NEVER logged, NEVER placed in a result field or
  evidence string, and NEVER in an error tuple. `verify.login` /
  `verify.studio` are anonymous/sentinel — they carry no secret. The route test
  scans every result byte to prove absence.

  ## Total over failure

  `run/1` NEVER raises and NEVER returns a bare transport error to the caller.
  An unreachable box is a first-class FAILED result: every probe reports
  `reachable: false`, `ok: false`, `status: nil`, and the envelope's `ok` /
  `reachable` are `false`. A hung box is bounded by the transport's per-request
  timeout (`BarkparkCloud.Verify.HttpClient`), so the whole synchronous suite is
  bounded (~15s worst case). The verdict is derived only from the probe
  outcomes.
  """

  alias BarkparkCloud.Registry
  alias BarkparkCloud.Registry.Barkpark

  # The deliberately-invalid sentinel credentials the login probe POSTs. Its ONLY
  # job is to make the auth/session stack ANSWER (401/422 = alive; 5xx = #957
  # dead-on-arrival). It authenticates nothing and carries no real secret —
  # byte-for-byte the intent of the Go gate's `verifyLoginSentinel`.
  @login_sentinel Jason.encode!(%{
                    "email" => "barkpark-verify-probe@invalid.example",
                    "password" => "deliberately-invalid-verify-sentinel"
                  })

  # ≤200 bytes of a response body ride into a failing probe's evidence (D53). A
  # chatty error page can never bloat the result. The per-probe HTTP bound that
  # keeps the synchronous suite ~15s worst case lives on the transport
  # (`BarkparkCloud.Verify.HttpClient`'s timeouts).
  @max_evidence_bytes 200

  # verify.studio follows the instance's scoped-login redirect. ≤3 hops matches
  # the Go health gate's `CheckStudio`; a chain longer than this never renders
  # and FAILS.
  @max_studio_hops 3

  # The probe vocabulary, as data — the single source `probes/0` exposes and the
  # fixture pins. `pass_rule` is the human-readable contract asserted byte-equal
  # against the fixture; the ACTUAL pass logic lives in `pass?/2` and is
  # exercised by real green/red probe tests, so the strings are never merely
  # decorative. Order is the gate order (api → login → studio).
  @probes [
    %{name: "verify.api", label: "API answers", pass_rule: "status == 200"},
    %{name: "verify.login", label: "Login responds", pass_rule: "status < 500"},
    %{
      name: "verify.studio",
      label: "Studio renders",
      pass_rule: "status < 500 after <= 3 redirect hops"
    }
  ]

  @type probe_result :: %{
          name: String.t(),
          ok: boolean(),
          reachable: boolean(),
          status: non_neg_integer() | nil,
          latency_ms: non_neg_integer(),
          evidence: String.t()
        }

  @type result :: %{
          ok: boolean(),
          reachable: boolean(),
          verified_at: String.t(),
          probes: [probe_result()]
        }

  @doc """
  The probe vocabulary (name / label / pass-rule), in gate order. The C8 test
  asserts this equals the decoded `verify_probes.json` fixture; the Go gate's
  compiled probe order is asserted against the same file.
  """
  @spec probes() :: [%{name: String.t(), label: String.t(), pass_rule: String.t()}]
  def probes, do: @probes

  @doc """
  Run the golden-path suite against a LIVE `barkpark` and return the result
  envelope.

    * `{:ok, result}` — the suite ran (all probes attempted). `result.ok` is the
      verdict; an unreachable box is a fully-populated `reachable: false`
      result, NOT an error.
    * `{:error, :not_live}` — no instance URL yet (still provisioning / a failed
      birth). Mirrors the instance-API proxy's liveness gate.
    * `{:error, :no_admin_token}` — the row never stored one (pre-feature box).
    * `{:error, :decrypt_failed}` — the stored ciphertext is tampered
      (fail-closed).

  The admin token never escapes this function.
  """
  @spec run(Barkpark.t()) ::
          {:ok, result()} | {:error, :not_live | :no_admin_token | :decrypt_failed}
  def run(%Barkpark{url: url} = bp) when is_binary(url) and url != "" do
    case Registry.reveal_admin_token(bp) do
      {:ok, nil} -> {:error, :no_admin_token}
      :error -> {:error, :decrypt_failed}
      {:ok, token} -> {:ok, execute(String.trim_trailing(url, "/"), token)}
    end
  end

  def run(_), do: {:error, :not_live}

  # Run the three probes in order and fold them into the envelope. Each probe is
  # self-contained (a transport error on one never aborts the others), so an
  # unreachable box yields three `reachable: false` probes rather than a raise.
  defp execute(base, token) do
    results = [
      probe_api(base, token),
      probe_login(base),
      probe_studio(base)
    ]

    %{
      ok: Enum.all?(results, & &1.ok),
      reachable: Enum.any?(results, & &1.reachable),
      verified_at: DateTime.to_iso8601(DateTime.utc_now()),
      probes: results
    }
  end

  # verify.api — authenticated GET /v1/capabilities must be 200. The admin token
  # rides ONLY here, and only in the request header (never the result).
  defp probe_api(base, token) do
    time_probe("verify.api", fn ->
      request(:get, base <> "/v1/capabilities", auth_headers(token), "")
    end)
    |> finish("verify.api", fn status -> status == 200 end, "GET /v1/capabilities → 200 (API up)")
  end

  # verify.login — sentinel wrong creds; PASS iff <500. A 401/422 proves the
  # auth stack answered; a 5xx is #957. Anonymous (no admin token).
  defp probe_login(base) do
    time_probe("verify.login", fn ->
      request(:post, base <> "/v1/auth/login", json_headers(), @login_sentinel)
    end)
    |> finish("verify.login", &(&1 < 500), fn status ->
      "POST /v1/auth/login → #{status} (auth stack answered; bad creds rejected)"
    end)
  end

  # verify.studio — GET /studio following the scoped-login redirect ≤3 hops;
  # final status <500 passes. Anonymous. Redirects are followed MANUALLY (the
  # transport does not auto-follow) so the ≤3-hop budget is enforced exactly,
  # matching the Go health gate's CheckStudio.
  defp probe_studio(base) do
    start = System.monotonic_time(:millisecond)
    outcome = follow_studio(base, base <> "/studio", @max_studio_hops)
    elapsed = System.monotonic_time(:millisecond) - start

    case outcome do
      {:ok, status, body} ->
        if status < 500 do
          probe("verify.studio", true, true, status, elapsed, "GET /studio → #{status} (renders)")
        else
          probe(
            "verify.studio",
            false,
            true,
            status,
            elapsed,
            "#{status} — #{body_snippet(body)}"
          )
        end

      :too_many_hops ->
        probe(
          "verify.studio",
          false,
          true,
          nil,
          elapsed,
          "redirect chain exceeded #{@max_studio_hops} hops (Studio never rendered)"
        )

      :unreachable ->
        unreachable_probe("verify.studio", elapsed)
    end
  end

  # Walk the redirect chain up to `hops` remaining. A 3xx with a Location is
  # followed (resolved relative to the current URL); anything else is the final
  # response. Running out of hops on a still-redirecting box FAILS.
  defp follow_studio(_base, _url, hops) when hops < 0, do: :too_many_hops

  defp follow_studio(base, url, hops) do
    case request(:get, url, [{"Accept", "text/html"}], "") do
      {:ok, %{status: status, body: body, headers: headers}} when status in 300..399 ->
        case redirect_location(headers) do
          nil ->
            {:ok, status, body}

          location ->
            follow_studio(base, resolve_url(base, url, location), hops - 1)
        end

      {:ok, %{status: status, body: body}} ->
        {:ok, status, body}

      {:error, _reason} ->
        :unreachable
    end
  end

  # ── probe plumbing ──

  # Time a single-request probe. Returns `{elapsed_ms, {:ok, resp} | {:error, _}}`.
  defp time_probe(_name, fun) do
    start = System.monotonic_time(:millisecond)
    result = fun.()
    {System.monotonic_time(:millisecond) - start, result}
  end

  # Fold a single-request probe's `{elapsed, transport_result}` into a
  # probe_result via a pass predicate + a success-evidence builder (a string, or
  # a fun of the status). A transport error is the reachable:false path.
  defp finish({elapsed, {:ok, %{status: status, body: body}}}, name, pass?, success) do
    if pass?.(status) do
      probe(name, true, true, status, elapsed, evidence(success, status))
    else
      probe(name, false, true, status, elapsed, "#{status} — #{body_snippet(body)}")
    end
  end

  defp finish({elapsed, {:error, _reason}}, name, _pass?, _success) do
    unreachable_probe(name, elapsed)
  end

  defp evidence(success, _status) when is_binary(success), do: success
  defp evidence(success, status) when is_function(success, 1), do: success.(status)

  defp probe(name, ok, reachable, status, elapsed, evidence) do
    %{
      name: name,
      ok: ok,
      reachable: reachable,
      status: status,
      latency_ms: max(elapsed, 0),
      evidence: truncate(evidence)
    }
  end

  # A box that never answered TCP — a FAILED probe, never a raise. No status, no
  # transport reason string (reasons can carry host detail; the honest signal is
  # simply "unreachable").
  defp unreachable_probe(name, elapsed) do
    probe(name, false, false, nil, elapsed, "instance unreachable")
  end

  # ── HTTP + header helpers ──

  defp auth_headers(token) do
    [
      {"Authorization", "Bearer " <> token},
      {"Accept", "application/json"}
    ]
  end

  defp json_headers do
    [
      {"Accept", "application/json"},
      {"Content-Type", "application/json"}
    ]
  end

  defp request(method, url, headers, body) do
    http_client().request(%{method: method, url: url, headers: headers, body: body})
  end

  # Case-insensitive Location lookup; a blank value is treated as absent.
  defp redirect_location(headers) do
    Enum.find_value(headers, fn {k, v} ->
      if String.downcase(k) == "location" and is_binary(v) and String.trim(v) != "", do: v
    end)
  end

  # Resolve a redirect target: an absolute URL is used verbatim; an
  # absolute-path ("/w/…") is re-based onto the instance origin; anything else
  # rides the current URL's directory. Instance scoped-login redirects are
  # absolute paths, the common case.
  defp resolve_url(base, _current, location) do
    cond do
      String.starts_with?(location, "http://") or String.starts_with?(location, "https://") ->
        location

      String.starts_with?(location, "/") ->
        base <> location

      true ->
        base <> "/" <> location
    end
  end

  # ── evidence hygiene ──

  # Truncate to ≤@max_evidence_bytes, trimmed, dropping any trailing bytes that
  # would split a UTF-8 codepoint (so the result is always valid JSON).
  defp truncate(text) when is_binary(text) do
    text
    |> byte_slice(@max_evidence_bytes)
    |> String.trim()
  end

  defp truncate(_), do: ""

  defp body_snippet(body) when is_binary(body),
    do: body |> byte_slice(@max_evidence_bytes) |> String.trim()

  defp body_snippet(_), do: ""

  defp byte_slice(bin, max) when byte_size(bin) <= max, do: bin

  defp byte_slice(bin, max) do
    sliced = binary_part(bin, 0, max)
    # Walk back off any partial trailing codepoint so the slice stays valid UTF-8.
    drop_partial_tail(sliced)
  end

  defp drop_partial_tail(bin) do
    if String.valid?(bin) do
      bin
    else
      case bin do
        <<>> -> <<>>
        _ -> drop_partial_tail(binary_part(bin, 0, byte_size(bin) - 1))
      end
    end
  end

  # The transport seam. Default `BarkparkCloud.Verify.HttpClient` (verified TLS
  # :httpc, returns response HEADERS so verify.studio can walk the redirect
  # chain); tests wire a header-aware fake.
  defp http_client do
    Application.get_env(:barkpark_cloud, :verify_http_client, BarkparkCloud.Verify.HttpClient)
  end
end

defmodule BarkparkCloud.Verify.HttpClient do
  @moduledoc """
  Verify's HTTP transport — verified-TLS `:httpc`, like
  `BarkparkCloud.Billing.HttpClient`, but it RETURNS response headers so
  `BarkparkCloud.Verify` can walk the `verify.studio` redirect chain (the
  billing client deliberately drops headers). Redirects are NOT auto-followed
  (`autoredirect: false`) — Verify follows them manually to enforce the exact
  ≤3-hop budget and to keep the SSRF-safe posture of the shared client.

  `request/1` takes the same request map shape and returns
  `{:ok, %{status, body, headers}}` on a completed exchange (any status) or
  `{:error, {:http_client, reason}}` on a transport failure. Header names are
  lower-cased so a case-insensitive `Location` lookup is trivial.
  """

  @timeout 5_000
  @connect_timeout 5_000

  @spec request(map()) ::
          {:ok, %{status: non_neg_integer(), body: binary(), headers: [{String.t(), String.t()}]}}
          | {:error, term()}
  def request(%{method: method, url: url, headers: headers} = req) do
    body = Map.get(req, :body, "")
    request_arg = to_httpc(method, url, headers, body)

    case :httpc.request(method, request_arg, http_opts(), body_format: :binary) do
      {:ok, {{_v, status, _reason}, resp_headers, resp_body}} ->
        {:ok,
         %{status: status, body: to_string(resp_body), headers: normalize_headers(resp_headers)}}

      {:error, reason} ->
        {:error, {:http_client, reason}}
    end
  end

  # GET carries no body; the others pass content-type + body pulled out (the
  # :httpc 4-tuple), mirroring the billing client.
  defp to_httpc(:get, url, headers, _body) do
    {to_charlist(url), header_charlists(headers)}
  end

  defp to_httpc(_method, url, headers, body) do
    {content_type, rest} = pop_content_type(headers)

    {to_charlist(url), header_charlists(rest), to_charlist(content_type), to_string(body)}
  end

  defp pop_content_type(headers) do
    {ct, rest} = Enum.split_with(headers, fn {k, _v} -> String.downcase(k) == "content-type" end)

    content_type =
      case ct do
        [{_k, v} | _] -> v
        [] -> "application/json"
      end

    {content_type, rest}
  end

  defp header_charlists(headers) do
    Enum.map(headers, fn {k, v} -> {to_charlist(k), to_charlist(v)} end)
  end

  defp normalize_headers(resp_headers) do
    Enum.map(resp_headers, fn {k, v} -> {String.downcase(to_string(k)), to_string(v)} end)
  end

  # Verified TLS against the OS trust store; redirects returned, not followed;
  # bounded timeouts so a hung instance can't wedge the synchronous suite.
  defp http_opts do
    [
      ssl: [
        verify: :verify_peer,
        cacerts: :public_key.cacerts_get(),
        customize_hostname_check: [
          match_fun: :public_key.pkix_verify_hostname_match_fun(:https)
        ]
      ],
      autoredirect: false,
      timeout: @timeout,
      connect_timeout: @connect_timeout
    ]
  end
end

File: internal/provisioner/verify.go
package provisioner

import (
	"context"
	"fmt"
	"io"
	"net/http"
	"strings"
	"time"

	"github.com/FRIKKern/barkpark/internal/cli/setup"
)

// C2 / D45 — the golden-path VERIFY gate.
//
// A box that passes the health gate is "healthy", but healthy is not the same as
// "the owner can actually create a server, open Studio, and log in". The #957
// class of failure — a 32-byte SECRET_KEY_BASE — leaves a box that answers
// /v1/capabilities and even 302s /studio, yet dies with a 500 the moment the
// session/cookie stack is exercised (login, the scoped Studio hop). The health
// gate's Studio check already walks the scoped redirect to catch part of this;
// VERIFY closes the loop by additionally proving the AUTH stack answers.
//
// The gate runs AFTER content bootstrap and BEFORE the provisioner declares the
// box ready, over HTTPS against the live instance origin. Four probes, in order:
//
//	verify.api       GET  /v1/capabilities        → 200                 (the API is up)
//	verify.login     POST /v1/auth/login (sentinel wrong creds) → <500  (auth answers)
//	verify.studio    GET  /studio (≤3 scoped hops) → final <500         (Studio renders)
//	verify.siteplane the site-hosting plane, WHEN REQUIRED               (sites can build)
//
// verify.siteplane is the one probe that is not an HTTP request: nothing the
// instance origin serves can see docker / nixpacks / the builder units. It reads
// a SITE-PLANE FACT handed to the gate by its caller (verifyConfig.sitePlane*)
// and is CONDITIONAL: a caller that does not require a plane gets a SKIPPED
// probe (it passes, and its evidence says it was skipped — never a fake
// "installed"). See verifySitePlane for who requires it and why.
//
// verify.login sends DELIBERATELY-WRONG sentinel credentials: a clean 401/422
// proves the whole request→session→auth pipeline ran and rejected them; ANY 5xx
// is the dead-on-arrival class and FAILS the gate. The minted admin token is
// held (for a future authenticated probe) but is NEVER sent by these anonymous /
// sentinel probes, and NEVER appears in any narrated evidence string.

const (
	// verifyProbeTimeout bounds each individual probe's HTTP request(s).
	verifyProbeTimeout = 10 * time.Second
	// verifyTotalBudget bounds the whole three-probe gate.
	verifyTotalBudget = 45 * time.Second
	// verifyMaxBodyBytes caps how much of a failing response body rides into the
	// failure evidence (≤200 bytes, per D45).
	verifyMaxBodyBytes = 200
)

// verifyLoginSentinel is the deliberately-invalid credential payload the login
// probe POSTs. It must never authenticate anything — its ONLY job is to make the
// auth/session stack answer (401/422 = alive; 5xx = #957 dead-on-arrival). It is
// NOT the minted admin token and carries no real secret.
const verifyLoginSentinel = `{"email":"barkpark-verify-probe@invalid.example","password":"deliberately-invalid-verify-sentinel"}`

// verifyConfig is one VERIFY gate invocation against a freshly provisioned box.
type verifyConfig struct {
	// baseURL is the instance origin (https://<label>.<zone> in prod; an httptest
	// fake instance in tests).
	baseURL string
	// token is the minted per-instance admin bearer. Held so an authenticated
	// probe could use it, but NEVER sent by the current anonymous/sentinel probes
	// and NEVER logged or narrated.
	token string
	// probeTimeout / totalBudget default to verifyProbeTimeout / verifyTotalBudget
	// when zero (production); tests set tiny values for fast fail paths.
	probeTimeout time.Duration
	totalBudget  time.Duration

	// sitePlaneRequired says whether THIS box must carry the site-hosting plane
	// for the gate to pass. False (the zero value) makes verify.siteplane a
	// SKIPPED probe — the restore path's setting, because a restored box is a
	// CMS resurrection whose plane nobody installed in this run, and a box that
	// legitimately has no plane must still restore green.
	sitePlaneRequired bool
	// sitePlaneComplete is the site-plane FACT the probe reads, three-state
	// under internal/agent/site_plane.go's law: nil UNMEASURED, false a measured
	// verdict that the plane is missing, true present. At birth it is the go-live
	// chain's own record of step 7c (cloud.LiveServer.SitePlaneInstalled).
	// Ignored when sitePlaneRequired is false.
	sitePlaneComplete *bool
	// sitePlaneMissing / sitePlaneUnmeasured / sitePlaneLogTail explain a false
	// sitePlaneComplete (cloud.LiveServer's diagnosis of the failed step 7c):
	// the components measured absent / not readable after the failed install,
	// and the last lines of the installer's own output. They ride into the
	// failure evidence so an upstream apt / nixpacks outage reads as one.
	sitePlaneMissing    []string
	sitePlaneUnmeasured []string
	sitePlaneLogTail    string
}

// probeOutcome is one probe's verdict plus the evidence + elapsed time narrated
// for it. evidence is a short human string — for a green probe a one-line
// success, for a red probe "<status> — <body ≤200 bytes>" or the transport
// error. It NEVER carries a secret.
type probeOutcome struct {
	name     string
	pass     bool
	evidence string
	elapsed  time.Duration
}

// runVerifyGate probes the golden path against a freshly provisioned instance and
// narrates each probe through report. Probes run in order; the FIRST red probe is
// reported as verify/failed and returned as an error so ProvisionWith fails the
// job (the box is torn down and the worker never POSTs /succeed). All green →
// verify/done and a nil error.
//
// The verdict is derived ONLY from the probe HTTP outcomes. report() tees to the
// StepReporter + live console and SWALLOWS their errors, so a dropped step/console
// report can never flip the gate — a red gate fails and a green gate passes
// regardless of telemetry delivery.
func runVerifyGate(ctx context.Context, cfg verifyConfig, report func(step, status, detail string)) error {
	if cfg.probeTimeout <= 0 {
		cfg.probeTimeout = verifyProbeTimeout
	}
	if cfg.totalBudget <= 0 {
		cfg.totalBudget = verifyTotalBudget
	}
	ctx, cancel := context.WithTimeout(ctx, cfg.totalBudget)
	defer cancel()

	// One client for the api + login probes: per-request timeout, and it does NOT
	// follow redirects (login's status is read verbatim). verify.studio reuses the
	// health gate's own hop-walking client via setup.HealthGate.CheckStudio.
	client := &http.Client{
		Timeout: cfg.probeTimeout,
		CheckRedirect: func(*http.Request, []*http.Request) error {
			return http.ErrUseLastResponse
		},
	}

	report("verify", "started", "")
	for _, probe := range verifyProbes {
		out := probe(ctx, cfg, client)
		if !out.pass {
			report("verify", "failed", fmt.Sprintf("%s: %s", out.name, out.evidence))
			return fmt.Errorf("golden-path %s: %s", out.name, out.evidence)
		}
		report("verify", "progress", fmt.Sprintf("%s: %s (%dms)", out.name, out.evidence, out.elapsed.Milliseconds()))
	}
	report("verify", "done", "")
	return nil
}

// verifyProbes is the gate's dispatch order — the ONE list runVerifyGate walks
// and TestProvisionerProbeVocabularyMatchesFixture holds to verify_probes.json.
// verify.siteplane rides LAST: it is the only probe that is not an HTTP read,
// and a box whose CMS is dead should be reported as that, not as a plane fault.
var verifyProbes = []func(context.Context, verifyConfig, *http.Client) probeOutcome{
	verifyAPI, verifyLogin, verifyStudio, verifySitePlane,
}

// verifyAPI (1): GET /v1/capabilities must return 200 — the API is up. Mirrors
// setup.HealthGate.checkCapabilities.
func verifyAPI(ctx context.Context, cfg verifyConfig, client *http.Client) probeOutcome {
	const name = "verify.api"
	start := time.Now()
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, cfg.baseURL+"/v1/capabilities", nil)
	if err != nil {
		return probeOutcome{name, false, "build request: " + err.Error(), time.Since(start)}
	}
	resp, err := client.Do(req)
	if err != nil {
		return probeOutcome{name, false, fmt.Sprintf("transport error: %v", err), time.Since(start)}
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return probeOutcome{name, false, fmt.Sprintf("%d — %s", resp.StatusCode, verifyBodySnippet(resp.Body)), time.Since(start)}
	}
	return probeOutcome{name, true, "GET /v1/capabilities → 200 (API up)", time.Since(start)}
}

// verifyLogin (2): POST /v1/auth/login with sentinel wrong creds. PASS iff the
// status is <500 — a 401/422 proves the request→session→auth pipeline ran and
// cleanly rejected the creds. Any 5xx is the #957 32-byte-SECRET_KEY_BASE
// dead-on-arrival class (the session/cookie stack crashes before auth) → FAIL.
func verifyLogin(ctx context.Context, cfg verifyConfig, client *http.Client) probeOutcome {
	const name = "verify.login"
	start := time.Now()
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, cfg.baseURL+"/v1/auth/login", strings.NewReader(verifyLoginSentinel))
	if err != nil {
		return probeOutcome{name, false, "build request: " + err.Error(), time.Since(start)}
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := client.Do(req)
	if err != nil {
		return probeOutcome{name, false, fmt.Sprintf("transport error: %v", err), time.Since(start)}
	}
	defer resp.Body.Close()
	if resp.StatusCode >= http.StatusInternalServerError {
		return probeOutcome{name, false, fmt.Sprintf("%d — %s", resp.StatusCode, verifyBodySnippet(resp.Body)), time.Since(start)}
	}
	return probeOutcome{name, true, fmt.Sprintf("POST /v1/auth/login → %d (auth stack answered; bad creds rejected)", resp.StatusCode), time.Since(start)}
}

// verifyStudio (3): GET /studio following the scoped-redirect chain (≤3 hops),
// final status <500 passes. It REUSES setup.HealthGate.CheckStudio so the exact
// hop-walking logic (and the reason it exists — a Studio that 302s then 500s one
// hop later) is shared, not duplicated. The context bounds the gate as a whole;
// the per-hop timeout is the health gate's own client timeout.
func verifyStudio(_ context.Context, cfg verifyConfig, _ *http.Client) probeOutcome {
	const name = "verify.studio"
	start := time.Now()
	hg := setup.HealthGate{
		BaseURL: cfg.baseURL,
		Token:   cfg.token, // held; CheckStudio is anonymous and does not send it
		Timeout: cfg.probeTimeout,
	}
	res := hg.CheckStudio()
	return probeOutcome{name, res.Pass, res.Detail, time.Since(start)}
}

// verifySitePlane (4): the site-hosting plane, CONDITIONALLY.
//
// Not required (sitePlaneRequired false) → PASS with evidence that says
// "skipped". That is the restore path and any caller with no expectation: a box
// that legitimately has no plane must not fail a gate over it.
//
// Required → PASS iff the fact is a measured true. A measured false FAILS (the
// box would go live with every site pointed at it sitting `queued`), and so does
// a nil: a required fact nobody measured is not a pass — a green with no
// subject is exactly the vacuous verdict the three-state law exists to refuse.
//
// ProvisionWith requires it iff the go-live chain actually ATTEMPTED step 7c
// (LiveServer.SitePlaneInstalled != nil — the control plane sent the agent
// token + control URL the plane installer rides on). An old control plane that
// never asked leaves it nil, and the probe is skipped: byte-for-byte the
// pre-probe verdict for those boxes.
func verifySitePlane(_ context.Context, cfg verifyConfig, _ *http.Client) probeOutcome {
	const name = "verify.siteplane"
	if !cfg.sitePlaneRequired {
		return probeOutcome{name, true, "skipped — no site plane required for this box", 0}
	}
	switch {
	case cfg.sitePlaneComplete == nil:
		return probeOutcome{name, false, "site plane required but UNMEASURED — nothing recorded whether it was installed", 0}
	case !*cfg.sitePlaneComplete:
		return probeOutcome{name, false, sitePlaneFailureEvidence(cfg), 0}
	default:
		return probeOutcome{name, true, "site plane installed (site-runtime installer exited 0)", 0}
	}
}

// sitePlaneFailureEvidence names what is missing and quotes the installer's
// last lines. Every clause is present only when it has content, but the
// "components" clause always says SOMETHING: when nothing was measured it says
// so, rather than implying a clean box.
func sitePlaneFailureEvidence(cfg verifyConfig) string {
	var b strings.Builder
	b.WriteString("site plane NOT installed — the site-runtime installer (go-live step 7c) failed; sites on this box would stay queued")
	if len(cfg.sitePlaneMissing) > 0 {
		b.WriteString("; missing: " + strings.Join(cfg.sitePlaneMissing, ", "))
	}
	if len(cfg.sitePlaneUnmeasured) > 0 {
		b.WriteString("; unmeasured: " + strings.Join(cfg.sitePlaneUnmeasured, ", "))
	}
	if len(cfg.sitePlaneMissing) == 0 && len(cfg.sitePlaneUnmeasured) == 0 {
		b.WriteString("; components: none recorded")
	}
	if cfg.sitePlaneLogTail != "" {
		b.WriteString("; step 7c log tail: " + cfg.sitePlaneLogTail)
	}
	return b.String()
}

// verifyBodySnippet reads at most verifyMaxBodyBytes of a failing response body
// for the failure evidence, trimmed. Bounded so a chatty error page can never
// bloat the narrated detail.
func verifyBodySnippet(r io.Reader) string {
	data, _ := io.ReadAll(io.LimitReader(r, verifyMaxBodyBytes))
	return strings.TrimSpace(string(data))
}
